qa/L0_lifecycle/lifecycle_test.py

```diff
@@ -2916,6 +2917,189 @@ def test_concurrent_load_unload(self):
+        for model in models:
+            # Instances should be loaded concurrently for supported backends
+            start_time = time.time()
+            try:
+                triton_client.load_model(model, config=json.dumps(config))
+            except Exception as ex:
+                self.assertTrue(False, "unexpected error {}".format(ex))
+            end_time = time.time()
+            loading_time = end_time - start_time
+            print(f"Time to load {num_instances} instances: {loading_time}")
+
+            # Each of the two models has a minimum loading delay of 10 seconds
+            # Speedup is observed when the concurrent loading time < 20 seconds
+            # but use a tighter bound of 15 seconds
+            self.assertLess(
+                loading_time, 15.0, "Concurrent loading speedup not observed"
+            )
+            # Concurrent loading time cannot be < 10 seconds
+            self.assertGreaterEqual(
+                loading_time, 10.0, "Invalid concurrent loading time"
+            )
+            # Make sure the models are loaded
+            self.assertTrue(triton_client.is_server_live())
+            self.assertTrue(triton_client.is_server_ready())
+            self.assertTrue(triton_client.is_model_ready(model))
+
+    def _call_with_timeout(self, callable, timeout_secs):
+        # Setup handler for timing out call
+        def timeout_handler(sig, frame):
+            raise TimeoutError()
+
+        signal.signal(signal.SIGALRM, timeout_handler)
+        signal.alarm(timeout_secs)
+        result = callable()
+        return result
+
+    def _call_with_expected_timeout(self, callable, timeout_secs=3):
+        # Call callable with expectation that it will timeout
+        try:
+            self._call_with_timeout(callable, timeout_secs)
+        except TimeoutError:
+            print("Inference timed out as expected.")
+            return
+        except Exception as ex:
+            self.assertTrue(False, "unexpected error {}".format(ex))
+        else:
+            self.assertTrue(False, "unexpected success, call should've timed out.")
+
+    def _get_fp32_io(self, client_type):
+        # Config
+        input_names = ["INPUT0", "INPUT1"]
+        output_names = ["OUTPUT0", "OUTPUT1"]
+        dtype, dims, shape = ("TYPE_FP32", [-1, 16], [1, 16])
+        input_config = [
+            {"name": name, "data_type": dtype, "dims": dims} for name in input_names
+        ]
+        output_config = [
+            {"name": name, "data_type": dtype, "dims": dims} for name in output_names
+        ]
+        # Inputs
+        inputs = []
+        for name in input_names:
+            inputs.append(
+                client_type.InferInput(name, shape, dtype.replace("TYPE_", ""))
+            )
+            inputs[-1].set_data_from_numpy(np.ones(shape, dtype=np.float32))
+        return input_config, output_config, inputs
+
+    def test_concurrent_model_instance_load_sanity(self):
```

Comment:
Just for my understanding. The goal of this test is to check if the included backends are capable of loading model instances concurrently?

Reply:
The goal is to check that nothing crashes and burns internally when trying to load instances in parallel for the supported backends. All it does it load N instances of various kinds, and ensure that exactly N instances are loaded (via sequence batch slot logic) and complete a successful inference (no exception is raised on the N inferences).

Reply:
Open to ideas if you think there's a better way to test

Reply:
I think the test case does not belong here, because the L0_lifecycle is for testing the correctness of the Triton core, while the correctness of the backend implementation belongs to the testing of each backend. Since the test is similar for all backends, maybe a new L0_backend_thread_safety (or something similar) can be added for this test?
@GuanLuo @tanmayv25 should correct me if you think otherwise

Reply:
@kthui do you have any feedback on the test itself other than how it is named? Or do you think it is sufficient for now?

Also @Tabrizian @tanmayv25 any feedback?

Reply:
I was wondering how is this testing the parallel instance feature? I see that it tests whether all the instances are ready or not but I don't see any testing in this specific case for parallel instance loading. I believe this test case would pass before this feature too. Please correct me if I'm missing anything.

Reply:
@Tabrizian you are correct that this subtest doesn't explicitly the parallelism in terms of speedup, I did that in the other subtest as it is easy to do for a python model, similar to the existing parallel model loading tests.

This sanity subtest was written as more of a sanity check that when loading a model with a significant (N=5, though upon writing I can probably increase this) instance count that (1) nothing crashes and burns, with higher instance counts it would be likely to reveal a race condition or segfault, hang, etc. if there was a bug, and (2) verify that exactly N instances (not more, not less, due to any possible race condition) do actually get loaded and can handle inference.

Upon reflection though, I can probably modify this test to do a serial load (env var) and then a parallel load and just check that the parallel load is faster at all. Initially I was afraid of this check introducing flakiness, but ideally it should consistently be faster on the types of CI systems we have with higher core counts.

Let me know if that makes sense, or if you have other ideas.